hooks/pre-exit

```diff
@@ -160,16 +161,55 @@ find_and_upload() {
       exit "${BUILDKITE_PLUGIN_TEST_COLLECTOR_MISSING_ERROR:-1}"
     fi
   else
+    declare -a uploads_in_progress=()
+    echo "Uploading '${#matching_files[@]}' files matching '${FILES_PATTERN}'"
+    
     # needs to be part of else for bash4.3 compatibility
     for file in "${matching_files[@]}"; do
-      echo "Uploading '$file'..."
-      if ! upload "$TOKEN_VALUE" "$FORMAT" "${file}"; then
-        echo "Error uploading, will continue"
-      fi
-      if [[ "$ANNOTATE" != "false" ]]; then
-        save-report-url "${CURL_RESP_FILE}"
-      fi
+      iterations_waited=0
+      while [[ "${#uploads_in_progress[@]}" -ge $UPLOAD_CONCURRENCY ]]; do
+        iterations_waited=$((iterations_waited + 1))
+        if [[ "${DEBUG}" == "true" ]]; then
+          echo "Waiting for uploads to finish..."
+        fi
+
+        sleep 1
+
+        for index in "${!uploads_in_progress[@]}"; do
+          # Note: kill -0 does not kill the pid, it provides a *nix compatible way to test the pid is responding.
+          if ! kill -0 "${uploads_in_progress[index]}" > /dev/null; then
+            unset 'uploads_in_progress[index]'
+          elif [[ "$iterations_waited" -gt $TIMEOUT ]]; then
+            echo "Upload '${uploads_in_progress[index]}' has been running for more than '${TIMEOUT}' seconds, killing it"
+            kill "${uploads_in_progress[index]}"
+            unset 'uploads_in_progress[index]'
+          fi
+        done
+      done
+
+      # Spawn a subcommand group to allow parallel uploads
+      {
+        echo "Uploading '$file'..."
+        
+        if ! upload "$TOKEN_VALUE" "$FORMAT" "${file}"; then
+          echo "Error uploading, will continue"
+        fi
+
+        if [[ "$ANNOTATE" != "false" ]]; then
+          save-report-url "${CURL_RESP_FILE}"
+        fi
```

Comment:
I just realized there is a bug here: the file is shared among all `upload` calls so there is a race condition as to which URL is saved for annotations and which are not. There are two alternatives to fix this:

1- make a note/check about annotation not compatible with concurrent uploads
2- change the `upload` function to accept another parameter that defines the filename to save the file that you can then pass through to the `save-report-url`

Whichever you prefer I'll be OK with

Reply:
Ah, good catch. I'll investigate this further.

Reply:
I think this is fixed now, I use `mktemp` to generate a temporary file in the command group if `CURL_RESP_FILE` is unset.

Reply:
I think this is fixed now, I use `mktemp` to generate a temporary file in the command group if `CURL_RESP_FILE` is unset.